File: Sample Apps/NodeJs Custom Attribute/otelapi/util/MotadataDynamicInstrumentation.js
const { trace } = require('@opentelemetry/api');

/**
 * Utility for adding custom attributes to OpenTelemetry spans.
 * All attribute keys are prefixed with "apm.".
 *
 * Supports only OpenTelemetry-compatible types:
 * string, number, boolean, string[], number[], boolean[]
 *
 * This class is thread-safe and exception-safe.
 * All methods silently ignore null values and will never throw exceptions.
 *
 * @since 1.0
 */
class MotadataDynamicInstrumentation {
  static DEFAULT_PREFIX = 'apm.';

  /**
   * Private helper to prefix keys with "apm."
   * @param {string} key - The attribute key
   * @returns {string|null} Prefixed key or null if invalid
   */
  static prefixKey(key) {
    try {
      if (key == null) {
        return null;
      }

      return key.startsWith(this.DEFAULT_PREFIX) ? key : this.DEFAULT_PREFIX + key;
    } catch (exception) {
      return null;
    }
  }

  /**
   * Core helper method to validate key and set attribute.
   * Executes the provided setter action if key is valid.
   *
   * @param {string} key - The attribute key
   * @param {*} value - The attribute value (for null check)
   * @param {Function} setter - The action to execute if validation passes
   */
  static safeSet(key, value, setter) {
    try {
      if (key == null || value == null) {
        return;
      }

      const prefixed = this.prefixKey(key);

      if (prefixed == null) {
        return;
      }

      // Pass prefixed key to setter to avoid calling prefixKey twice
      setter(prefixed);
    } catch (exception) {
      // Ignore all exceptions
    }
  }

  /**
   * Sets a String, Number, or Boolean attribute on the current span.
   *
   * @param {string} key - The attribute key
   * @param {string|number|boolean} value - The attribute value
   */
  static set(key, value) {
    // Handle string type
    if (typeof value === 'string') {
      this.safeSet(key, value, (prefixed) => {
        const span = trace.getActiveSpan();
        if (span) {
          span.setAttribute(prefixed, value);
        }
      });
    }
    // Handle number type
    else if (typeof value === 'number') {
      this.safeSet(key, value, (prefixed) => {
        const span = trace.getActiveSpan();
        if (span) {
          span.setAttribute(prefixed, value);
        }
      });
    }
    // Handle boolean type
    else if (typeof value === 'boolean') {
      this.safeSet(key, value, (prefixed) => {
        const span = trace.getActiveSpan();
        if (span) {
          span.setAttribute(prefixed, value);
        }
      });
    }
  }

  /**
   * Sets a List of Boolean values attribute on the current span.
   *
   * @param {string} key - The attribute key
   * @param {boolean[]} value - The list of boolean values
   */
  static setBooleanList(key, value) {
    try {
      if (key == null || value == null || !Array.isArray(value) || value.length === 0) {
        return;
      }

      const prefixed = this.prefixKey(key);

      if (prefixed == null) {
        return;
      }

      const span = trace.getActiveSpan();
      if (span) {
        span.setAttribute(prefixed, value);
      }
    } catch (exception) {
      // Ignore all exceptions
    }
  }

  /**
   * Sets a List of Number values attribute on the current span.
   *
   * @param {string} key - The attribute key
   * @param {number[]} value - The list of number values
   */
  static setNumberList(key, value) {
    try {
      if (key == null || value == null || !Array.isArray(value) || value.length === 0) {
        return;
      }

      const prefixed = this.prefixKey(key);

      if (prefixed == null) {
        return;
      }

      const span = trace.getActiveSpan();
      if (span) {
        span.setAttribute(prefixed, value);
      }
    } catch (exception) {
      // Ignore all exceptions
    }
  }

  /**
   * Sets a List of String values attribute on the current span.
   *
   * @param {string} key - The attribute key
   * @param {string[]} value - The list of string values
   */
  static setStringList(key, value) {
    try {
      if (key == null || value == null || !Array.isArray(value) || value.length === 0) {
        return;
      }

      const prefixed = this.prefixKey(key);

      if (prefixed == null) {
        return;
      }

      const span = trace.getActiveSpan();
      if (span) {
        span.setAttribute(prefixed, value);
      }
    } catch (exception) {
      // Ignore all exceptions
    }
  }
}

module.exports = MotadataDynamicInstrumentation;

File: Sample Apps/Java Custom Attribute/src/OtelApi/util/MotadataDynamicInstrumentation.java
package OtelApi.util;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.common.AttributeKey;
import java.util.List;
import java.util.ArrayList;

/**
 * Utility for adding custom attributes to OpenTelemetry spans.
 * All attribute keys are prefixed with "apm.".
 *
 * <p>Supports only OpenTelemetry-compatible types:
 * boolean, Boolean, double, Double, int, Integer, long, Long, String,
 * List&lt;Boolean&gt;, List&lt;Double&gt;, List&lt;Integer&gt;, List&lt;Long&gt;, List&lt;String&gt;
 *
 * <p>This class is thread-safe and exception-safe.
 * All methods silently ignore null values and will never throw exceptions.
 *
 * @since 1.0
 */
public final class MotadataDynamicInstrumentation
{

    private static final String DEFAULT_PREFIX = "apm.";

    private MotadataDynamicInstrumentation()
    {
    }

    private static String prefixKey(String key)
    {
        try
        {
            if (key == null)
            {
                return null;
            }

            return key.startsWith(DEFAULT_PREFIX) ? key : DEFAULT_PREFIX + key;
        }
        catch (Exception exception)
        {
            return null;
        }
    }

    /**
     * Core helper method to validate key and set attribute.
     * Executes the provided setter action if key is valid.
     *
     * @param key the attribute key
     * @param value the attribute value (for null check)
     * @param setter the action to execute if validation passes
     */
    private static void safeSet(String key, Object value, Runnable setter)
    {
        try
        {
            if (key == null || value == null)
            {
                return;
            }

            String prefixed = prefixKey(key);

            if (prefixed == null)
            {
                return;
            }

            setter.run();

        }
        catch (Exception exception)
        {
            // Ignore all exceptions
        }
    }

    /**
     * Sets a Boolean attribute on the current span.
     * Accepts both primitive boolean and Boolean wrapper.
     *
     * @param key the attribute key
     * @param value the attribute value
     */
    public static void set(String key, Boolean value)
    {
        safeSet(key, value, () ->
        {
            String prefixed = prefixKey(key);

            Span.current().setAttribute(prefixed, value);
        });
    }

    /**
     * Sets a Double attribute on the current span.
     * Accepts both primitive double and Double wrapper.
     *
     * @param key the attribute key
     * @param value the attribute value
     */
    public static void set(String key, Double value)
    {
        safeSet(key, value, () ->
        {
            String prefixed = prefixKey(key);

            Span.current().setAttribute(prefixed, value);
        });
    }

    /**
     * Sets an Integer attribute on the current span.
     * Accepts both primitive int and Integer wrapper.
     *
     * @param key the attribute key
     * @param value the attribute value
     */
    public static void set(String key, Integer value)
    {
        safeSet(key, value, () ->
        {
            String prefixed = prefixKey(key);

            Span.current().setAttribute(prefixed, value.longValue());
        });
    }

    /**
     * Sets a Long attribute on the current span.
     * Accepts both primitive long and Long wrapper.
     *
     * @param key the attribute key
     * @param value the attribute value
     */
    public static void set(String key, Long value)
    {
        safeSet(key, value, () ->
        {
            String prefixed = prefixKey(key);

            Span.current().setAttribute(prefixed, value);
        });
    }

    /**
     * Sets a String attribute on the current span.
     *
     * @param key the attribute key
     * @param value the attribute value
     */
    public static void set(String key, String value)
    {
        safeSet(key, value, () ->
        {
            String prefixed = prefixKey(key);

            Span.current().setAttribute(prefixed, value);
        });
    }

    /**
     * Sets a List&lt;Boolean&gt; attribute on the current span.
     *
     * @param key the attribute key
     * @param value the list of boolean values
     */
    public static void setBooleanList(String key, List<Boolean> value)
    {
        try
        {
            if (key == null || value == null || value.isEmpty())
            {
                return;
            }

            String prefixed = prefixKey(key);

            if (prefixed == null)
            {
                return;
            }

            Span.current().setAttribute(AttributeKey.booleanArrayKey(prefixed), value);

        }
        catch (Exception exception)
        {
            // Ignore all exceptions
        }
    }

    /**
     * Sets a List&lt;Double&gt; attribute on the current span.
     *
     * @param key the attribute key
     * @param value the list of double values
     */
    public static void setDoubleList(String key, List<Double> value)
    {
        try
        {
            if (key == null || value == null || value.isEmpty())
            {
                return;
            }

            String prefixed = prefixKey(key);

            if (prefixed == null)
            {
                return;
            }

            Span.current().setAttribute(AttributeKey.doubleArrayKey(prefixed), value);

        }
        catch (Exception exception)
        {
            // Ignore all exceptions
        }
    }

    /**
     * Sets a List&lt;Integer&gt; attribute on the current span.
     *
     * @param key the attribute key
     * @param value the list of integer values
     */
    public static void setIntegerList(String key, List<Integer> value)
    {
        try
        {
            if (key == null || value == null || value.isEmpty())
            {
                return;
            }

            String prefixed = prefixKey(key);

            if (prefixed == null)
            {
                return;
            }

            // Convert Integer to Long for OpenTelemetry compatibility
            List<Long> longList = new ArrayList<>(value.size());

            for (Integer item : value)
            {
                if (item != null)
                {
                    longList.add(item.longValue());
                }
            }

            if (!longList.isEmpty())
            {
                Span.current().setAttribute(AttributeKey.longArrayKey(prefixed), longList);
            }

        }
        catch (Exception exception)
        {
            // Ignore all exceptions
        }
    }

    /**
     * Sets a List&lt;Long&gt; attribute on the current span.
     *
     * @param key the attribute key
     * @param value the list of long values
     */
    public static void setLongList(String key, List<Long> value)
    {
        try
        {
            if (key == null || value == null || value.isEmpty())
            {
                return;
            }

            String prefixed = prefixKey(key);

            if (prefixed == null)
            {
                return;
            }

            Span.current().setAttribute(AttributeKey.longArrayKey(prefixed), value);

        }
        catch (Exception exception)
        {
            // Ignore all exceptions
        }
    }

    /**
     * Sets a List&lt;String&gt; attribute on the current span.
     *
     * @param key the attribute key
     * @param value the list of string values
     */
    public static void setStringList(String key, List<String> value)
    {
        try
        {
            if (key == null || value == null || value.isEmpty())
            {
                return;
            }

            String prefixed = prefixKey(key);

            if (prefixed == null)
            {
                return;
            }

            Span.current().setAttribute(AttributeKey.stringArrayKey(prefixed), value);

        }
        catch (Exception exception)
        {
            // Ignore all exceptions
        }
    }
}